caffe2/core/dispatch/KernelRegistration.h

```diff
@@ -0,0 +1,137 @@
+#pragma once
+
+#include "caffe2/core/dispatch/OpSchema.h"
+#include "caffe2/core/dispatch/Dispatcher.h"
+#include "caffe2/utils/Optional.h"
+
+/**
+ * To register your own kernel for an operator, do in one (!) cpp file:
+ *   C10_REGISTER_KERNEL(OpSchemaDef)
+ *      .kernel(&kernel_func)
+ *      .dispatchKey(dispatch_key);
+ */
+
+namespace c10 {
+
+// TODO Test different order for builder
+// TODO Test no dispatch key defined
+
+/**
+ * Class which, on construction, registers an operator in the dispatch table.  The intent is that
+ * this class is constructed at static initialization time so that operators automatically get
+ * registered when a dlopen() occurs.
+ *
+ * You shouldn't call this directly; instead, use the KernelRegistrationBuilder
+ *
+ * @tparam OpSchemaDef
+ */
+template<class OpSchemaDef>
+class KernelRegistrar final {
+private:
+    using Schema = OpSchema<OpSchemaDef>;
+public:
+  /**
+   * @param kernel The concrete function implementation to register
+   * @param dispatch_key  The dispatch key to register the function to
+   */
+  KernelRegistrar(typename Schema::signature::func_type* kernel, typename Schema::dispatch::dispatch_key_type dispatch_key)
+  : dispatch_key_(std::move(dispatch_key)), owns_registration_(true) {
+    Dispatcher<OpSchemaDef>::registerOp(kernel, dispatch_key_);
+  }
+
+  KernelRegistrar(KernelRegistrar&& rhs)
+  : dispatch_key_(std::move(rhs.dispatch_key_)), owns_registration_(true) {
+    rhs.owns_registration_ = false;
+  }
+
+  // not needed for now
+  KernelRegistrar& operator=(KernelRegistrar&& rhs) = delete;
+
+  ~KernelRegistrar() {
+    if (owns_registration_) {
+      Dispatcher<OpSchemaDef>::deregisterOp(dispatch_key_);
+    }
+  }
+
+private:
+  const typename Schema::dispatch::dispatch_key_type dispatch_key_;
+  bool owns_registration_;
+
+  DISABLE_COPY_AND_ASSIGN(KernelRegistrar);
+};
+
+/**
+ * Helper class for building a KernelRegistrar.  This permits "keyword-argument" like syntax
+ * when performing operator registration, e.g., as in:
+ *
+ * C10_REGISTER_KERNEL(::ops::add_notensor)
+ *      .kernel(&add_notensor_op)
+ *      .dispatchKey("bla");
+ *
+ * Expanded, this macro invocation looks like:
+ *
+ * static KernelRegistrar<::ops::add_notensor> _anon0 =
+ *    KernelRegistrationBuilder<::ops::add_notensor, false, false>()
+ *      .kernel(&add_notensor_op)
+ *      .dispatchKey("bla");
+ *
+ * The resulting full expression is implicitly convertible to a KernelRegistrar.
+ *
+ * @tparam OpSchemaDef The operator schema this is building a KernelRegistration for
+ * @tparam hasKernel Boolean for compile-time checking that a kernel is specified before finalizing the builder
+ * @tparam hasDispatchKey Boolean for compile-time checking thhat a dispatch key is specified before finalizing the builder
+ */
+template<class OpSchemaDef, bool hasKernel, bool hasDispatchKey>
```

Comment:
it's a not very scalable approach to add every field in the  template args.

Also, why is constexpr important here? Since it's static registration, this code would be much more readable if just written usual regular fields (i.e. fill out the class and when it's initialized in the variable, check that everything is set).

Reply:
The goal is to move the builder pattern entirely to compile time. One reason is to have a low static initialization time overhead, the other reason is to get compile time errors instead of runtime errors if you forgot to give it a dispatch key or kernel function. I'm hoping we don't need too many arguments for kernel registration, so scalability wouldn't be an issue.

Reply:
There's an alternative solution that makes it more extensible, but it's a whole lot more of metaprogramming and I don't think it's worth it.

The basic idea would be to accumulate the data in the builder pattern by having a inheritance chain where each class adds one data field (say the kernel function) to the fields of its parent class, and whenever you call a function on the builder, it builds a inheritance chain of these classes and has all the fields. If we really get too many arguments, we can consider it. For now, I don't think it's worth the added complexity.

Reply:
What's bad about runtime static initialization error? Anybody writing an op would need to test/run it, so the error would be spotted right away.

It just feels like quite a complication. E.g. if you have 8-10 fields like in https://github.com/pytorch/pytorch/blob/master/caffe2/core/operator_schema.h it'd be a bit annoying to manipulate.

One option for the builder pattern is to template on mask instead of separate bool fields, at least that'd be a bit more readable:
```
template<class OpSchemaDef, int fields_mask=0>
class KernelRegistrationBuilder final {
...
constexpr KernelRegistrationBuilder<OpSchemaDef, fields_mask | FIELD_KERNEL> kernel(typename Schema::signature::func_type* kernel_func) && {
+    static_assert(!(fields_mask & FIELD_KERNEL), "Tried to define kernel twice in same op registration");
+    return KernelRegistrationBuilder<OpSchemaDef, fields_mask | FIELD_KERNEL>(*kernel_func, std::move(dispatch_key_));
+  }
```

Reply:
Sounds good, changing to a flags parameter.